sys/net/gnrc/link_layer/lwmac/tx_state_machine.c

```diff
@@ -0,0 +1,457 @@
+#define LOG_INFO(...) LOG(LOG_INFO, "[lwmac-tx] " __VA_ARGS__)
+#define LOG_DEBUG(...) LOG(LOG_DEBUG, "[lwmac-tx] " __VA_ARGS__)
+
+/* Break out of switch and mark the need for rescheduling */
+#define GOTO_TX_STATE(tx_state, do_resched) lwmac->tx.state = tx_state; \
+                                reschedule = do_resched; \
+                                break
+
+/******************************************************************************/
+
+void lwmac_tx_start(lwmac_t* lwmac, gnrc_pktsnip_t* pkt, lwmac_tx_neighbour_t* neighbour)
+{
+    assert(lwmac != NULL);
+    assert(pkt != NULL);
+    assert(neighbour != NULL);
+
+    if(lwmac->tx.packet) {
+        LOG_WARNING("Starting but tx.packet is still set\n");
+        gnrc_pktbuf_release(lwmac->tx.packet);
+    }
+
+    lwmac->tx.packet = pkt;
+    lwmac->tx.current_neighbour = neighbour;
+    lwmac->tx.state = TX_STATE_INIT;
+    lwmac->tx.wr_sent = 0;
+}
+
+/******************************************************************************/
+
+void lwmac_tx_stop(lwmac_t* lwmac)
+{
+    if(!lwmac)
+        return;
+
+    lwmac_clear_timeout(lwmac, TIMEOUT_WR);
+    lwmac_clear_timeout(lwmac, TIMEOUT_NO_RESPONSE);
+    lwmac_clear_timeout(lwmac, TIMEOUT_NEXT_BROADCAST);
+    lwmac_clear_timeout(lwmac, TIMEOUT_BROADCAST_END);
+    lwmac->tx.state = TX_STATE_STOPPED;
+
+    /* Release packet in case of failure */
+    if(lwmac->tx.packet) {
+        gnrc_pktbuf_release(lwmac->tx.packet);
+        lwmac->tx.packet = NULL;
+    }
+    lwmac->tx.current_neighbour = NULL;
+}
+
+/******************************************************************************/
+
+/* Returns whether rescheduling is needed or not */
+static bool _lwmac_tx_update(lwmac_t* lwmac)
+{
+    bool reschedule = false;
+
+    if(!lwmac)
+        return reschedule;
+
+    switch(lwmac->tx.state)
+    {
+    case TX_STATE_INIT:
+    {
+        lwmac_clear_timeout(lwmac, TIMEOUT_WR);
+        lwmac_clear_timeout(lwmac, TIMEOUT_NO_RESPONSE);
+        lwmac_clear_timeout(lwmac, TIMEOUT_NEXT_BROADCAST);
+        lwmac_clear_timeout(lwmac, TIMEOUT_BROADCAST_END);
+
+        if(_packet_is_broadcast(lwmac->tx.packet)) {
+            /* Set CSMA retries as configured and enable */
+            uint8_t csma_retries = LWMAC_BROADCAST_CSMA_RETRIES;
+			lwmac->netdev2_driver->set(lwmac->netdev->dev, NETOPT_CSMA_RETRIES,
+                                        &csma_retries, sizeof(csma_retries));
+            netopt_enable_t csma_enable = NETOPT_ENABLE;
+			lwmac->netdev2_driver->set(lwmac->netdev->dev, NETOPT_CSMA, &csma_enable, sizeof(csma_enable));
+
+            GOTO_TX_STATE(TX_STATE_SEND_BROADCAST, true);
+        } else {
+            /* Don't attempt to send a WR if channel is busy to get timings
+             * right, will be changed for sending DATA packet */
+            netopt_enable_t csma_disable = NETOPT_DISABLE;
+			lwmac->netdev2_driver->set(lwmac->netdev->dev, NETOPT_CSMA, &csma_disable, sizeof(csma_disable));
```

Comment:
Yes, that's true. But using CSMA was no option either since it introduces a random wait interval before sending for the first time. So yes, it should regard it as a normal transmission failure if I remember correctly.